pepper-apis/src/main/java/org/broadinstitute/ddp/service/AddressService.java

```diff
@@ -281,6 +282,23 @@ public boolean deleteAddress(Handle handle, String addressGuid) {
         return findDefaultAddressForParticipant(handle, participantGuid, DEFAULT_OLC_PRECISION);
     }
 
+    public Optional<MailAddress> findDefaultAddressForParticipantWithRetries(Handle handle, String participantGuid, int retries) {
+        do {
+            Optional<MailAddress> optionalAddress = findDefaultAddressForParticipant(handle, participantGuid);
+            if (optionalAddress.isPresent()) {
+                return optionalAddress;
+            } else {
+                try {
+                    LOG.warn("Could not find default address for participant with guid {}. Waiting to retry", participantGuid);
```

Comment:
Will be interesting to scan the logs to see how often retries were actually necessary.

Reply:
This approach works with transaction isolation level READ COMMITTED or lower.
I checked that in dev it is READ COMMITTED. Hopefully on production is the same level.

Reply:
That is a good point. We set the isolation level in the JDBC URL and when setting connection pool.
```
jdbc:mysql:///pepperapisprod?cloudSqlInstance=XXXXXX&socketFactory=com.google.cloud.sql.mysql.SocketFactory&user=XXXXXX&password=XXXXXX&characterEncoding=UTF-8&useLegacyDatetimeCode=false&serverTimezone=UTC&useSSL=false&sessionVariables=innodb_strict_mode=on,tx_isolation='READ-COMMITTED',sql_mode='TRADITIONAL'
```
and from our Java code:

https://github.com/broadinstitute/ddp-study-server/blob/fde8856a299b8f4525e883cc73225d16ada4af5c/pepper-apis/src/main/java/org/broadinstitute/ddp/db/TransactionWrapper.java#L381

Presumably the JDBC will set the default for everything and the Java code sets is up for the specific connection. So I think we have it cover.